drivers/ad5770r/ad5770r.c

```diff
@@ -0,0 +1,650 @@
+ */
+int32_t ad5770r_set_monitor_setup(struct ad5770r_dev *dev,
+				  struct monitor_setup mon_setup)
+{
+	int32_t ret;
+	ret = ad5770r_spi_reg_write(dev,
+				    AD5770R_MONITOR_SETUP,
+				    AD5770R_MONITOR_SETUP_MON_CH(mon_setup.monitor_channel) |
+				    AD5770R_MONITOR_SETUP_IB_EXT_EN(mon_setup.ib_ext_en) |
+				    AD5770R_MONITOR_SETUP_MUX_BUFFER(mon_setup.mux_buffer) |
+				    AD5770R_MONITOR_SETUP_MON_FUNCTION(mon_setup.monitor_function));
+
+	if (ret)
+		return -1;
+
+	dev->mon_setup = mon_setup;
+
+	return ret;
+};
+
+/**
+ * Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ *		       parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_init(struct ad5770r_dev **device,
+		     struct ad5770r_init_param init_param)
+{
+	struct ad5770r_dev	*dev;
+	uint8_t product_id_l, product_id_h;
+	int32_t ret;
+	enum ad5770r_channels i;
+
+	dev = (struct ad5770r_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return -1;
+
+	/* SPI */
+	ret = spi_init(&dev->spi_desc, init_param.spi_init);
+
+	/* Query device presence */
+	ad5770r_spi_reg_read(dev, AD5770R_PRODUCT_ID_L, &product_id_l);
+	ad5770r_spi_reg_read(dev, AD5770R_PRODUCT_ID_H, &product_id_h);
+
+	if (product_id_l != 0x04 || product_id_h != 0x40) {
+		printf("failed to read id (0x%X : 0x%X)\n", product_id_l,
+		       product_id_h);
+		return -1;
+	}
+
+	/* Device settings */
+	ret = ad5770r_set_device_spi(dev,
+				     init_param.dev_spi_settings);
+
+	ret |= ad5770r_channel_config(dev,
+				      init_param.channel_config);
+	ret |= ad5770r_set_channel_en(dev,
+				      init_param.channel_enable);
+
+	ret |= ad5770r_set_reference(dev,
+				     init_param.external_reference,
+				     init_param.reference_selector);
+	ret |= ad5770r_set_alarm(dev, init_param.alarm_config);
+
+	ret |= ad5770r_set_hw_ladc(dev,
+				   init_param.mask_hw_ldac);
+
+	for( i = AD5770R_CH0; i <=  AD5770R_CH5; i++) {
+		ret |= ad5770r_set_output_mode(dev,
+					       init_param.output_mode[i], i);
+		ret |= ad5770r_set_output_filter(dev,
+						 init_param.output_filter[i], i);
+		ret |= ad5770r_set_dac_value(dev,
+					     init_param.dac_value[i], i);
+		ret |= ad5770r_set_dac_input(dev,
+					     init_param.input_value[i], i);
+	}
+
+	dev->output_mode = init_param.output_mode;
+	dev->output_filter = init_param.output_filter;
+	dev->dac_value = init_param.dac_value;
+	dev->input_value = init_param.input_value;
+
+	ret |= ad5770r_set_page_mask(dev,
+				     init_param.page_mask);
+	ret |= ad5770r_set_mask_channel(dev,
+					init_param.mask_channel_sel);
+	ret |= ad5770r_set_sw_ldac(dev,
+				   init_param.sw_ldac);
+	ret |= ad5770r_set_monitor_setup(dev,
+					 init_param.mon_setup);
+
+	*device = dev;
+
+	if (!ret)
+		printf("ad5770r successfully initialized\n");
+	else
+		printf("ad5770r initialization error (%d)\n", ret);
+
+	mdelay(1000);
```

Comment:
maybe remove this delay; some users may not want it; or would like to add their own delay